libs/core/langchain_core/messages/utils.py

```diff
@@ -242,3 +255,650 @@ def convert_to_messages(
+
+            from langchain_core.messages import trim_messages, AIMessage, BaseMessage, HumanMessage, SystemMessage
+
+            messages = [
+                SystemMessage("This is a 4 token text. The full message is 10 tokens."),
+                HumanMessage("This is a 4 token text. The full message is 10 tokens.", id="first"),
+                AIMessage(
+                    [
+                        {"type": "text", "text": "This is the FIRST 4 token block."},
+                        {"type": "text", "text": "This is the SECOND 4 token block."},
+                    ],
+                    id="second",
+                ),
+                HumanMessage("This is a 4 token text. The full message is 10 tokens.", id="third"),
+                AIMessage("This is a 4 token text. The full message is 10 tokens.", id="fourth"),
+            ]
+
+            def dummy_token_counter(messages: List[BaseMessage]) -> int:
+                # treat each message like it adds 3 default tokens at the beginning
+                # of the message and at the end of the message. 3 + 4 + 3 = 10 tokens
+                # per message.
+
+                default_content_len = 4
+                default_msg_prefix_len = 3
+                default_msg_suffix_len = 3
+
+                count = 0
+                for msg in messages:
+                    if isinstance(msg.content, str):
+                        count += default_msg_prefix_len + default_content_len + default_msg_suffix_len
+                    if isinstance(msg.content, list):
+                        count += default_msg_prefix_len + len(msg.content) *  default_content_len + default_msg_suffix_len
+                return count
+
+        First 30 tokens, not allowing partial messages:
+        .. code-block:: python
+
+            trim_messages(messages, n_tokens=30, token_counter=dummy_token_counter, strategy="first")
+
+        .. code-block:: python
+
+            [
+                SystemMessage("This is a 4 token text. The full message is 10 tokens."),
+                HumanMessage("This is a 4 token text. The full message is 10 tokens.", id="first"),
+            ]
+
+        First 30 tokens, allowing partial messages:
+        .. code-block:: python
+
+            trim_messages(
+                messages,
+                n_tokens=30,
+                token_counter=dummy_token_counter,
+                strategy="first",
+                allow_partial=True,
+            )
+
+        .. code-block:: python
+
+            [
+                SystemMessage("This is a 4 token text. The full message is 10 tokens."),
+                HumanMessage("This is a 4 token text. The full message is 10 tokens.", id="first"),
+                AIMessage( [{"type": "text", "text": "This is the FIRST 4 token block."}], id="second"),
+            ]
+
+        First 30 tokens, allowing partial messages, have to end on HumanMessage:
+        .. code-block:: python
+
+            trim_messages(
```

Comment:
not sure if you'd agree but seeing `messages` as a value for each block will make it easier to see what the API does